go/ethadapter/blockprovider.go

```diff
@@ -0,0 +1,272 @@
+		// So check manually:
+		block := e.ethClient.FetchHeadBlock()
+		return h == block.Hash()
+	}
+	return h == e.liveBlocks.latest.Hash()
+}
+
+// streamBlocks should be run in a separate go routine. It will stream catch-up blocks from requested height until it
+// reaches the latest live block, then it will block until next live block arrives
+// It blocks when:
+// - publishing a block, it blocks on the outbound channel until the block is consumed
+// - awaiting a live block, when consumer is completely up-to-date it waits for a live block to arrive
+func (e *EthBlockProvider) streamBlocks() {
+	atomic.StoreInt32(e.runningStatus, int32(statusCodeRunning))
+	for !e.stopped() {
+		// this will block if we're up-to-date with live blocks
+		block, err := e.nextBlockToStream()
+		if err != nil {
+			e.logger.Error("unexpected error while preparing block to stream, will retry in 1 sec", log.ErrKey, err)
+			time.Sleep(time.Second)
+			continue
+		}
+		e.logger.Trace("blockProvider streaming block", "height", block.Number(), "hash", block.Hash())
+		e.streamCh <- block // we block here until consumer takes it
+		// update stream state
+		e.latestSent = block.Header()
+	}
+}
+
+func (e *EthBlockProvider) nextBlockToStream() (*types.Block, error) {
+	if e.latestSent == nil {
+		blk, err := e.ethClient.BlockByNumber(e.streamFrom)
+		if err != nil {
+			return nil, fmt.Errorf("no block at requested height=%s - %w", e.streamFrom, err)
+		}
+
+		return blk, nil
+	}
+
+	head, err := e.liveBlocks.AwaitNewBlock(context.TODO(), e.latestSent.Hash())
+	if err != nil {
+		return nil, fmt.Errorf("no new block found from stream - %w", err)
+	}
+
+	// most common path should be: new head block that arrived is the next block, and needs to be sent
+	if head.ParentHash == e.latestSent.Hash() {
+		blk, err := e.ethClient.BlockByHash(head.Hash())
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch block with hash=%s - %w", head.Hash(), err)
+		}
+		return blk, nil
+	}
+
+	// and if not then, we find the latest canonical block we sent and try one after that
+	latestCanon, err := e.latestCanonAncestor(e.latestSent.Hash())
+	if err != nil {
+		return nil, fmt.Errorf("could not find ancestor on canonical chain for hash=%s - %w", e.latestSent.Hash(), err)
+	}
+	// and send the cannon block after the fork
+	blk, err := e.ethClient.BlockByNumber(increment(latestCanon.Number()))
+	if err != nil {
+		return nil, fmt.Errorf("could not find block after canon fork branch, height=%s - %w", increment(latestCanon.Number()), err)
+	}
+	return blk, nil
+}
+
+// checkStopped checks the status for stopped code
+func (e *EthBlockProvider) stopped() bool {
+	return atomic.LoadInt32(e.runningStatus) == int32(statusCodeStopped)
+}
+
+// LiveBlocksMonitor manages a process that queues up the latest X blocks (where X=queueCap) in a FIFO queue, streamed from
+// an L1 client. If it has X and receives another one it will drop the oldest.
+type LiveBlocksMonitor struct {
```

Comment:
I struggle a bit to understand the responsibility of this object. Is the comment accurate?

Is this a good summary:
 This type abstracts away the L1 block subscription, which can be:

1. A wrapper over a geth subscription that handles reconnects
2. Something that polls for blocks in case the subscription is not supported ( like our mock stuff)

Reply:
If it's something that can have a clear responsibility that can be stated, then it deserves an abstraction.
Otherwise, probably best to fold back into the block provider. 
if it turns out there's some good abstraction, we can always extract it